crates/rustc_utils/src/cache.rs

```diff
@@ -1,56 +1,147 @@
+//! 
+//! To avoid this scenario both caches are implemented to detect when a
+//! recursive call as described is performed and `get` will panic. If your code
+//! uses recursive construction and would like to handle this case gracefully
+//! use [`get_maybe_recursive`](Cache::get_maybe_recursive) instead wich returns
+//! `None` from `get(k)` *iff* `k` this call (potentially transitively)
+//! originates from another `get(k)` call.
+//! 
+//! [^inconsistent]: For any given cache value `get` should only ever be used
+//!     with one, referentially transparent `compute` function. Essentially this
+//!     means running `compute(k)` should always return the same value
+//!     *independent of the state of it's environment*. Violation of this rule
+//!     can introduces non-determinism in your program.
+use std::{cell::RefCell, hash::Hash, pin::Pin};
 
 use rustc_data_structures::fx::FxHashMap as HashMap;
 
 /// Cache for non-copyable types.
-pub struct Cache<In, Out>(RefCell<HashMap<In, Pin<Box<Out>>>>);
+pub struct Cache<In, Out>(RefCell<HashMap<In, Option<Pin<Box<Out>>>>>);
 
 impl<In, Out> Cache<In, Out>
 where
   In: Hash + Eq + Clone,
 {
+  /// Size of the cache
+  pub fn len(&self) -> usize {
+    self.0.borrow().len()
+  }
   /// Returns the cached value for the given key, or runs `compute` if
   /// the value is not in cache.
+  /// 
+  /// # Panics
+  /// 
+  /// Returns `None` if this is a recursive invocation of `get` for key `key`.
   pub fn get<'a>(&'a self, key: In, compute: impl FnOnce(In) -> Out) -> &'a Out {
+    self.get_maybe_recursive(key, compute).unwrap_or_else(recursion_panic)
+  }
+  /// Returns the cached value for the given key, or runs `compute` if
+  /// the value is not in cache.
+  /// 
+  /// Returns `None` if this is a recursive invocation of `get` for key `key`.
+  pub fn get_maybe_recursive<'a>(&'a self, key: In, compute: impl FnOnce(In) -> Out) -> Option<&'a Out> {
     if !self.0.borrow().contains_key(&key) {
+      self.0.borrow_mut().insert(key.clone(), None);
       let out = Box::pin(compute(key.clone()));
-      self.0.borrow_mut().insert(key.clone(), out);
+      self.0.borrow_mut().insert(key.clone(), Some(out));
     }
 
     let cache = self.0.borrow();
-    let entry_pin = cache.get(&key).unwrap();
-    let entry_ref = entry_pin.as_ref().get_ref();
+    // Important here to first `unwrap` the `Option` created by `get`, then
+    // propagate the potential option stored in the map.
+    let entry = cache.get(&key).expect("invariant broken").as_ref()?;
 
     // SAFETY: because the entry is pinned, it cannot move and this pointer will
     // only be invalidated if Cache is dropped. The returned reference has a lifetime
     // equal to Cache, so Cache cannot be dropped before this reference goes out of scope.
-    unsafe { mem::transmute::<&'_ Out, &'a Out>(entry_ref) }
+    Some(unsafe { std::mem::transmute::<&'_ Out, &'a Out>(&**entry) })
   }
 }
 
+fn recursion_panic<A>() -> A {
```

Comment:
`fn recursion_panic() -> !` would be more direct.

Reply:
I had that originally, but that meant I could not pass the function pointer to `unwrap_or_else`. I fell like this is a weakness in rust's type checking.

Reply:
If you feel very strongly about this I can change it, but I'll have to make it `.unwrap_or_else(|| recursion_panic())` everywhere which seems kinda ugly

Reply:
Ah, I assumed that Rust would treat it equivalently to `<A>() -> A`. That’s a little sad. We can keep it as is.

Reply:
I thought that too 😞